### ISSUE TITLE
i18n support

### ISSUE THREAD

**Issue by luismanson** [HemmeligOrg/Hemmelig.app#78]
Hi, I'm the guy who asked about translation support in Reddit. Would you consider a PR with some support? Frontend maybe.. I think I can add some basic implementation of react-i18n...

**Comment by bjarneo** [comment]
👋 

Please do. Hopfully the frontend code is not to spaghetti. I know strings are thrown around all the places. Even with ternary operators.

react-i18n is 👍 to me!

**Comment by luismanson** [comment]
Hey, What do you think? https://github.com/luismanson/Hemmelig.app/commit/563ffbf7cca9c49f56fd3cfdb1f9522cb0a6def8
it's not very complete, but the basic front strings.

**Comment by bjarneo** [comment]
Hey @luismanson, I will look into it today :)

**Comment by bjarneo** [comment]
@luismanson added a comment: https://github.com/luismanson/Hemmelig.app/commit/563ffbf7cca9c49f56fd3cfdb1f9522cb0a6def8#commitcomment-83183511

To me it looks great. You do not have to add the loader component for the Suspense, but if you do it would be cool. Feel free to open a PR :)

**Comment by bjarneo** [comment]
PR: https://github.com/HemmeligOrg/Hemmelig.app/pull/82

**Comment by bjarneo** [comment]
new PR: https://github.com/HemmeligOrg/Hemmelig.app/pull/83

Thank you for your contribution :)

**Comment by RainerZufahl** [comment]
How would I go about adding more languages? I've created a new file `public/locales/de/translation.json` with the updated strings. Do I need to add this new language in `src/i18n.js` as well? And does the site itself already have some sort of system language detection or would that still be needed?

**Comment by luismanson** [comment]
Hello, you need to set `de` as the default language in `src/i18n.js`. 
Language switch/detection has not implemented yet.

PS: @bjarneo I'll submit my translation in the next day or so.

**Comment by RainerZufahl** [comment]
If i can figure out at what point the strings are loaded, maybe I can whip up a language detection & automatic switching method. Wish me luck 🤞🏼

**Comment by luismanson** [comment]
check this doc: https://react.i18next.com/latest/using-with-hooks the example has some comments about language detection and file loading

**Comment by bjarneo** [comment]
Reopening this issue.

One thing to consider here is to have an option to set the default language for the instance. This can be solved by adding an env var called `SECRET_LANGUAGE`. Then add than env var to the client config here: https://github.com/HemmeligOrg/Hemmelig.app/blob/main/config/default.cjs#L63

Which means you can fetch it like this: https://github.com/HemmeligOrg/Hemmelig.app/blob/main/src/client/routes/home/index.js#L51
One important note here is that the config that is imported and used in the frontend, is coming from this file: https://github.com/HemmeligOrg/Hemmelig.app/blob/main/src/client/config.js

**Comment by bjarneo** [comment]
Btw, so cool that you guys want to contribute 🎉

**Comment by bjarneo** [comment]
I have added you both @luismanson and @RainerZufahl as contributors to this repository. Let me know if something is not right about the settings. Should be allowed to create branches and PRs

**Comment by RainerZufahl** [comment]
Thank you! I've already created a fork of the repo which I'm working in, should I scrap that and create a new branch in this repo?

**Comment by bjarneo** [comment]
@rainerzufahl that is up to you, will work either way

**Comment by luismanson** [comment]
Wow, thanks! I'll test if works with the spanish json later today.

**Comment by RainerZufahl** [comment]
automatic language detection is now present via #88.

**Comment by luismanson** [comment]
How should we handle new strings? I think this has to be as transparent as possible. 
I noticed later that new strings could be written as `t('some_string', 'Some String')` but there are other options as [Runtime Extraction](https://www.i18next.com/how-to/extracting-translations#2-using-an-extraction-tool) and some [plugins](https://www.i18next.com/overview/plugins-and-utils).

**Comment by bjarneo** [comment]
Hm. Interesting. I have to look into this.

But, of course, using t('some.string', 'Default string') is a good start, because that means if a language does not have the translation, it will automatically fallback to the default english string?

**Comment by luismanson** [comment]
Yes, it should. If i understand correctly, this would even work with an empty default LANG setting.

**Comment by bjarneo** [comment]
I think that would be the first approach for this. How would runtime extraction and plugins work with create-react-app?

**Comment by luismanson** [comment]
I'll have to look into it, i do not know tbh.

**Comment by RainerZufahl** [comment]
> But, of course, using t('some.string', 'Default string') is a good start, because that means if a language does not have the translation, it will automatically fallback to the default english string?

That would mean that the english translations are hardcoded into the program, instead of an external string file, correct? While practical in some ways, that would make adjusting the english strings later on more complicated again.

**Comment by ltguillaume** [comment]
Continuing from #212

- Currently, `SECRET_FORCED_LANGUAGE=??` does not result in the language changing.
- It's fine that there's an option to force the language, but imho this should not be the default behavior. Instead:
    1. Don't set `SECRET_FORCED_LANGUAGE` by default in `docker-compose.yml`. If set, use the defined locale.
    2. Use the _browser's_ preferred language list to determine the language
    3. Fall back to English

**Comment by bjarneo** [comment]
@ltguillaume this PR solves the SECRET_FORCED_LANGUAGE issue: https://github.com/HemmeligOrg/Hemmelig.app/commit/1404f0ac58e82f4e57b275aa0cbda57d89d6f26a

**Comment by bjarneo** [comment]
Yeah, that would've been something

**Comment by bjarneo** [comment]
No, I have not looked into this.

**Comment by bjarneo** [comment]
<img width="817" alt="Screenshot 2024-01-22 at 20 49 37" src="https://github.com/HemmeligOrg/Hemmelig.app/assets/1419214/55223479-08dc-418b-8528-a69458cdd6de">
https://github.com/HemmeligOrg/Hemmelig.app/commit/1a9d84e5de509ad46fdcef3ef9b32029225159cb

new release coming soon with this working based off of the browser language

**Comment by bjarneo** [comment]
This is in production right now:
<img width="1229" alt="Screenshot 2024-01-22 at 21 02 50" src="https://github.com/HemmeligOrg/Hemmelig.app/assets/1419214/1ee77574-3847-4613-bcc1-85cf6a829f64">

**Comment by bjarneo** [comment]
@ltguillaume no worries. Let me know if it does not work as intended. :)

**Comment by bjarneo** [comment]
Yes, that was it. Do not be harsh on yourself. It still has been implemented in a way not needed for Hemmelig as far as I can tell. 

As for your bullet points, that is something that has to be looked into. 

1. Not sure. Have to look into this
2. Then it should be added as translatable strings
3. Should be an fairly easy fix
4. This should be refactored to return status codes, and let the frontend handle the strings, which again is translateable

**Comment by bjarneo** [comment]
Right, that makes sense. It should be an easy fix, I do not know the fix, but it should exclude /signin. Probably just a setting somewhere.

**Comment by bjarneo** [comment]
Yeah, and this application do not support this kind of routing (yet), so I am unsure if it has ever worked for these routes.

**Comment by bjarneo** [comment]
Yes, I was thinking about removing this, but forgot to try to do so while fixing the navigator. Removed it now, and looks like that solved it: https://github.com/HemmeligOrg/Hemmelig.app/commit/be6ba9e202ca2dc832052f83258090f9bee2db18
Creating a new release. Great catch!

**Comment by bjarneo** [comment]
@ltguillaume [v5.16.6](https://github.com/HemmeligOrg/Hemmelig.app/releases/tag/v5.16.6)

This contains the removal of path.
